### PR TITLE
Remove custom Snapdrop-Fork

### PR DESCRIPTION
**Pull request by ueen**
Integrate https://github.com/fm-sys/snapdrop in the app

so i added
- copy to JSI
- change tweet
- added settings button
- handle serverconnection-active/updateLastOnlineTime 
- handle recipient-shortclicked/shouldOpenSendTextDialog
- remove donations button
- remove weird sound (safari hack)

please tell me if there are other important changes (what about `getVersionId`?)

Tested in Firefox with Tampermonkey. it should be executed at `document-idle` (should be fine with `onPageFinished`)

Also added onResume `refreshWebsite` behaviour guarded by `onlinePastThreeMin` to prevent refreshing to soon or while picking a file

### CODE REVIEW

**Comment by fm-sys** [comment]
This. Is. Great! Thank you very much ❤️
 
Originally I wanted to do it like that, but didn't know how to overwrite the js methods. Haven't tested it yet, but this seems to be the best solution, because we don't need to deal with backward compatibility. Only thing is that we **never ever** are allowed to display an external donation link (requirement of Google Play) so we need to hope that the website structure will not change 😅

**Comment by ueen** [comment]
Tested it and fixed an quote in quote in quote escape error :D
Seems to work just fine, please check it out!
There seems to be a few random sounds, did you do something about that?
Ah the donation, yea i'll remove that (tbh Google is not that thorough if it comes to review)

**Comment by fm-sys** [comment]
> There seems to be a few random sounds, did you do something about that?

Yes, I remember that. These lines are the problem:

https://github.com/fm-sys/snapdrop/blob/b54dfa7d001bfea5fd2d8bdbd4b69e9c8113f6b0/client/scripts/ui.js#L618

**Comment by fm-sys** [comment]
Ok, compiled the code and tested it on an Android  10 emulator. At a first glance everything seems to be fine. If everything is done from your side, I will try some edge cases, which where making trouble in the past, and merge afterwards.

**Comment by ueen** [comment]
Great! Please do test, I have another PR lined up https://github.com/ueen/snapdrop-android/pull/1to integrate my initial intention, adding onResume and check for transfer before refreshing. Do you want to test this first? Let me know if everything works I merge, you test again and then you can merge?

**Comment by fm-sys** [comment]
> please tell me if there are other important changes (what about `getVersionId`?)

I think you caught everything. 
`getVersionId` is currently unused. I had recently implemented it to be able to read out the app version in the future. Probably with the latest changes it is now obsolete...

File: app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java [hunk 1]
```diff
@@ -230,18 +230,67 @@ public static String getSendTextDialogWithPreInsertedString(final String text) {
...
+                //change ReceiveTextDialog._onCopy to connect to JavaScriptInterface
+                "ReceiveTextDialog.prototype._oC = ReceiveTextDialog.prototype._onCopy;" +
+                "ReceiveTextDialog.prototype._onCopy = function(){" +
+                "               this._oC();" +
+                "               SnapdropAndroid.copyToClipboard(this.$text.textContent);" +
+                "            };" +
+                //change PeerUI.setProgress(progress) to connect to JavaScriptInterface
+                "PeerUI.prototype.sP = PeerUI.prototype.setProgress;" +
+                "PeerUI.prototype.setProgress = function(progress){" +
+                "               SnapdropAndroid.setProgress(progress);" +
```
Comment (fm-sys) [inline app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java:242]: Crashes currently, as it's not yet implemented

File: app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java [hunk 2]
```diff
@@ -230,18 +230,67 @@ public static String getSendTextDialogWithPreInsertedString(final String text) {
...
+                "                   SnapdropAndroid.updateLastOnlineTime();" +
+                "               }" +
+                "            };" +
+                
+                //change PeerUI._onTouchEnd(e) to connect to JavaScriptInterface
+                "PeerUI.prototype._oTE = PeerUI.prototype._onTouchEnd;" +
+                "PeerUI.prototype._onTouchEnd = function(e){" +
+                "               this._oTE(e);" +
+                "               if ((Date.now() - this._touchStart < 500) && SnapdropAndroid.shouldOpenSendTextDialog()) {" +
+                "                   Events.fire('text-recipient', e.detail);" +
```
Comment (fm-sys) [inline app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java:288]: Somehow, the send text dialog opens, but sending doesn't work when coming from shortclick. Need to dive into it

**Review by fm-sys — COMMENTED** [review]
Ok, unfortunately some things showed up. But I think it's easier if I debug them in android studio directly. So I would merge both of your PR's and tackle the remaining issues afterwards (as soon as time permits). Ok? 

BTW, it was to late for me yesterday so I stopped testing. Sorry.

File: app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java [hunk 3]
```diff
@@ -230,18 +230,67 @@ public static String getSendTextDialogWithPreInsertedString(final String text) {
     }
     public static String initialiseWebsite() {
         return "javascript: " +
+                //change ReceiveTextDialog._onCopy to connect to JavaScriptInterface
+                "ReceiveTextDialog.prototype._oC = ReceiveTextDialog.prototype._onCopy;" +
+                "ReceiveTextDialog.prototype._onCopy = function(){" +
+                "               this._oC();" +
```
[the remark this answers is not in the document]
Comment (ueen) [reply app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java:236]: i think an Event is fired as well, in my tests it works just fine, can you provide a Stacktrace?

File: app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java (hunk 2 above)
Comment (ueen) [reply app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java:288]: Ah i see, thats strange....

File: app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java (hunk 1 above)
Comment (ueen) [reply app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java:242]: yea true! i merged my patch-2 to fix

**Comment by ueen** [comment]
Ok i merged the other one into this one as well but removed onResume for now because it has some nasty sideeffects like reloading after picking a file which breaks sending a file from the app

File: app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java (hunk 2 above)
Comment (ueen) [reply app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java:288]: Oh i think i was tired as well, of course it needs to be `this._peer.id` its fixed now

**Comment by fm-sys** [comment]
Thanks again for this very great contribution!!!

**Comment by ueen** [comment]
Will you release the update to the playstore soon? :)

**Comment by fm-sys** [comment]
Hope so. I'm on vacation currently, but maybe I will find time at some point in the next few days

**Comment by ueen** [comment]
Oh enjoy your free days 😊

File: app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fm-sys) [reply app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java:236]: > The js maybe throws an expection because navigator isnt accessible from inside the WebView, but this should be a silent one; filtering by chromium in Logcat i get Uncaught (in promise) NotAllowedError: Write permission denied. but everything works :)

OK, copy works indeed. Just the optical feedback (Snackbar "Copied to clipboard") is missing. 

> I think we should touch the js as little as possible (add code instead of overriding) not to break anything should there be minor changes to the website

Generally agree. But I think that in case of `_onCopy()` it makes sense to overwrite the code (because we don't want to copy twice ;-) I think there's not much to be expected which could get broken in the future.)

[the remark this answers is not in the document]
Comment (ueen) [reply app/src/main/java/com/fmsys/snapdrop/JavaScriptInterface.java:236]: I think navigator doesn't work, in webview, so it isn't copied twice.
The snackbar I don't know, but you can add it in the JSI :)

**Comment by fm-sys** [comment]
Sorry, it took a bit longer until I found time to work a bit on *Snapdrop for Android* and prepare a new release. Unfortunately, everything takes a little more time than it seems at first glance :-/

Anyway, release is created and uploaded to google play. Presumably it will be accepted and rolled out sometime in the next few hours...
